packages/core/src/fhir-deduplication/resources/condition.ts

```diff
@@ -0,0 +1,114 @@
+import { CodeableConcept, Condition } from "@medplum/fhirtypes";
+import {
+  ICD_10_CODE,
+  ICD_10_OID,
+  SNOMED_CODE,
+  SNOMED_OID,
+  combineTwoResources,
+  createCompositeKey,
+  getDateFromString,
+} from "../shared";
+
+/**
+ * Approach:
+ * 1. Group same Conditions based on:
+ *      - Medical codes:
+ *          - ICD-10, if possible
+ *          // TODO: Introduce SNOMED cross-walk to match SNOMED with ICD-10
+ *          - SNOMED, if possible
+ *      - Date
+ * 2. Combine the Conditions in each group into one master condition and return the array of only unique and maximally filled out Conditions
+ */
+export function deduplicateConditions(conditions: Condition[]) {
+  const { snomedMap, icd10Map, remainingConditions } = groupSameConditions(conditions);
+  return combineConditions(snomedMap, icd10Map, remainingConditions);
+}
+
+export function groupSameConditions(conditions: Condition[]): {
+  icd10Map: Map<string, Condition>;
+  snomedMap: Map<string, Condition>;
+  remainingConditions: Condition[];
+} {
+  const snomedMap = new Map<string, Condition>();
+  const icd10Map = new Map<string, Condition>();
+  const remainingConditions: Condition[] = [];
+
+  for (const condition of conditions) {
+    const date = createDateKey(condition);
+    const { snomedCode, icd10Code } = extractCodes(condition.code);
+
+    if (icd10Code) {
+      const compKey = JSON.stringify(createCompositeKey(icd10Code, date));
```

Comment:
if date is undefined, the compKey will be 
"code | undefined"

which means we will dedup two of the same codes on two undefined dates. Is this desired behavior? I think this either
1. should not be desired
2. the date getting logic should be modified. There are a bunch of date fields we dont look at. If we actually check all date fields and both conditions have nothing, then they should be deduped as two dateless conditions. 

```
  onsetDateTime?: string;
  onsetAge?: Age;
  onsetPeriod?: Period;
  onsetRange?: Range;
  onsetString?: string;
  abatementDateTime?: string;
  abatementAge?: Age;
  abatementPeriod?: Period;
  abatementRange?: Range;
  abatementString?: string;
  recordedDate?: string;
  ```
  Either way, we should make this explicit with a test that captures this behavior. Ditto on snomed kv store

Reply:
I looked at the FHIR converter code, and we're not using any of these other fields anywhere, so we won't encounter them during deduplication either. 

If two conditions are the same and neither has a date, then they might as well be grouped together and deduplicated imo

Reply:
Ok if thats the case, can you write a test for that behavior? Imo its unintuitive behavior that should be captured explicitly by a test

Reply:
The only other concern would be if we change the fhir converter. Just because we dont do it know doesn't mean we wont add it. At a high level, from now on, whenever we make changes to the FHIR converter, we will need to make sure to look at the deduplication logic and make sure it accounts for those changes. No action item here now but just bringing it up!